src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -352,42 +352,31 @@ public BigInteger(ReadOnlySpan<byte> value, bool isUnsigned = false, bool isBigE
             }
             else
             {
-                int unalignedBytes = byteCount % 4;
-                int dwordCount = byteCount / 4 + (unalignedBytes == 0 ? 0 : 1);
-                uint[] val = new uint[dwordCount];
-                int byteCountMinus1 = byteCount - 1;
+                int wholeDwordCount = Math.DivRem(byteCount, 4, out int unalignedBytes);
+                uint[] val = new uint[wholeDwordCount + (unalignedBytes == 0 ? 0 : 1)];
 
                 // Copy all dwords, except don't do the last one if it's not a full four bytes
-                int curDword, curByte;
-
                 if (isBigEndian)
                 {
-                    curByte = byteCount - sizeof(int);
-                    for (curDword = 0; curDword < dwordCount - (unalignedBytes == 0 ? 0 : 1); curDword++)
+                    int curByte = byteCount - sizeof(int);
+                    for (int curDword = 0; curDword < wholeDwordCount; curDword++)
                     {
-                        for (int byteInDword = 0; byteInDword < 4; byteInDword++)
-                        {
-                            byte curByteValue = value[curByte];
-                            val[curDword] = (val[curDword] << 8) | curByteValue;
-                            curByte++;
-                        }
+                        val[curDword] = BinaryPrimitives.ReadUInt32BigEndian(value.Slice(curByte));
 
-                        curByte -= 8;
+                        curByte -= 4;
                     }
                 }
                 else
                 {
-                    curByte = sizeof(int) - 1;
-                    for (curDword = 0; curDword < dwordCount - (unalignedBytes == 0 ? 0 : 1); curDword++)
-                    {
-                        for (int byteInDword = 0; byteInDword < 4; byteInDword++)
-                        {
-                            byte curByteValue = value[curByte];
-                            val[curDword] = (val[curDword] << 8) | curByteValue;
-                            curByte--;
-                        }
+                    ReadOnlySpan<uint> dwords = MemoryMarshal.Cast<byte, uint>(value.Slice(0, wholeDwordCount * 4));
```

Comment:
This is not a safe operation for the BCL to make since the input isn't guaranteed to be aligned.

Reply:
Oh. Thanks. When would this show up in tests?

Are any of the following different?

https://github.com/dotnet/runtime/blob/8c8b40b462a9629718ca8859f8a109ebc0bf28cb/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs#L384

https://github.com/dotnet/runtime/blob/8c8b40b462a9629718ca8859f8a109ebc0bf28cb/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/OpenSslX509ChainProcessor.cs#L294

https://github.com/dotnet/runtime/blob/8c8b40b462a9629718ca8859f8a109ebc0bf28cb/src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs#L341

https://github.com/dotnet/runtime/blob/8c8b40b462a9629718ca8859f8a109ebc0bf28cb/src/libraries/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceCounterLib.cs#L128-L129

Reply:
Arrays are aligned, while arbitrary spans may not.

Reply:
Understood, thanks. I will revert this bit